pulsar-client-cpp/lib/ClientConnection.cc

```diff
@@ -114,13 +114,16 @@ isTlsAllowInsecureConnection_(false) {
     if (clientConfiguration.isUseTls()) {
         using namespace boost::filesystem;
 
-        boost::asio::ssl::context ctx(boost::asio::ssl::context::tlsv12_client);
-
+#if BOOST_VERSION >= 105400
+        boost::asio::ssl::context ctx(executor_->io_service_, boost::asio::ssl::context::tlsv12_client);
```

Comment:
I am still confused on why we need to pass the `io_service` here, in particular the one from the executor which is kind of unrelated to the connection itself.

As for the macro, I'd prefer to have it defined in the top section of the file. Something like
```c++
#if BOOST_VERSION >= 105400
   #define TLS_CLIENT = boost::asio::ssl::context::tlsv12_client
#else
   #define TLS_CLIENT = boost::asio::ssl::context::tlsv1_client
#endif
```